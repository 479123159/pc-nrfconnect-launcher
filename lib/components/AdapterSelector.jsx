/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React, { PropTypes } from 'react';
import { DropdownButton } from 'react-bootstrap';
import { Iterable } from 'immutable';
import AdapterSelectorItem from './AdapterSelectorItem';

class AdapterSelector extends React.Component {
    componentDidMount() {
        const {
            bindHotkey,
            toggleExpanded,
            hotkeyExpand,
        } = this.props;

        bindHotkey(hotkeyExpand.toLowerCase(), toggleExpanded);
    }

    renderAdapterItems() {
        const {
            adapters,
            onSelect,
            isLoading,
        } = this.props;

        if (!isLoading) {
            return adapters.map(adapter => (
                <AdapterSelectorItem
                    key={adapter.comName}
                    adapter={adapter}
                    onSelect={onSelect}
                />
            ));
        }
        return null;
    }

    renderCloseItem() {
        const {
            selectedAdapter,
            onDeselect,
        } = this.props;

        if (selectedAdapter) {
            return (
                <AdapterSelectorItem
                    key="closeAdapter"
                    adapter={{
                        comName: 'Close adapter',
                        serialNumber: '',
                    }}
                    onSelect={onDeselect}
                />
            );
        }
        return null;
    }

    render() {
        const {
            selectedAdapter,
            adapterIndicator,
            toggleExpanded,
            isExpanded,
            hotkeyExpand,
        } = this.props;

        const selectorText = selectedAdapter || 'Select serial port';

        return (
            <span title={`Select serial port (${hotkeyExpand})`}>
                <div className="padded-row">
                    <DropdownButton
                        id="navbar-dropdown"
                        className="btn-primary btn-nordic"
                        title={selectorText}
                        open={isExpanded}
                        onToggle={toggleExpanded}
                    >
                        { this.renderAdapterItems() }
                        { this.renderCloseItem() }
                    </DropdownButton>
                    <div className={`indicator ${adapterIndicator}`} />
                </div>
            </span>
        );
    }
}

AdapterSelector.propTypes = {
    adapters: PropTypes.oneOfType([
        PropTypes.instanceOf(Array),
        PropTypes.instanceOf(Iterable),
    ]).isRequired,
    selectedAdapter: PropTypes.string,
    adapterIndicator: PropTypes.string,
    isLoading: PropTypes.bool,
    isExpanded: PropTypes.bool,
    toggleExpanded: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
    onDeselect: PropTypes.func.isRequired,
    bindHotkey: PropTypes.func.isRequired,
    hotkeyExpand: PropTypes.string,
};

AdapterSelector.defaultProps = {
    selectedAdapter: '',
    isExpanded: false,
    isLoading: false,
    adapterIndicator: 'off',
    hotkeyExpand: 'Alt+P',
};

export default AdapterSelector;
