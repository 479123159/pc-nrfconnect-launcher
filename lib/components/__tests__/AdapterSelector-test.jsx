/* eslint-disable import/first */

jest.mock('react-bootstrap', () => ({
    DropdownButton: 'DropdownButton',
    MenuItem: 'MenuItem',
}));

import React from 'react';
import renderer from 'react-test-renderer';
import AdapterSelector from '../AdapterSelector';

describe('AdapterSelector', () => {
    it('should render empty adapter list, not expanded', () => {
        expect(renderer.create(
            <AdapterSelector
                adapters={[]}
                toggleExpanded={() => {}}
                onSelect={() => {}}
                onDeselect={() => {}}
                bindHotkey={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render empty adapter list, expanded', () => {
        expect(renderer.create(
            <AdapterSelector
                adapters={[]}
                isExpanded
                toggleExpanded={() => {}}
                onSelect={() => {}}
                onDeselect={() => {}}
                bindHotkey={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render two adapters', () => {
        expect(renderer.create(
            <AdapterSelector
                adapters={[
                    {
                        comName: '/dev/tty0',
                        serialNumber: '123456',
                    }, {
                        comName: '/dev/tty1',
                        serialNumber: '456789',
                    },
                ]}
                isExpanded
                toggleExpanded={() => {}}
                onSelect={() => {}}
                onDeselect={() => {}}
                bindHotkey={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render two adapters, one selected', () => {
        expect(renderer.create(
            <AdapterSelector
                adapters={[
                    {
                        comName: '/dev/tty0',
                        serialNumber: '123456',
                    }, {
                        comName: '/dev/tty1',
                        serialNumber: '456789',
                    },
                ]}
                selectedAdapter={'/dev/tty1'}
                isExpanded
                toggleExpanded={() => {}}
                onSelect={() => {}}
                onDeselect={() => {}}
                bindHotkey={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render empty list while loading', () => {
        expect(renderer.create(
            <AdapterSelector
                adapters={[
                    {
                        comName: '/dev/tty0',
                        serialNumber: '123456',
                    },
                ]}
                isExpanded
                isLoading
                toggleExpanded={() => {}}
                onSelect={() => {}}
                onDeselect={() => {}}
                bindHotkey={() => {}}
            />,
        )).toMatchSnapshot();
    });
});
