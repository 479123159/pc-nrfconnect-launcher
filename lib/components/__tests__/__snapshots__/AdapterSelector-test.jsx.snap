exports[`AdapterSelector should render empty adapter list, expanded 1`] = `
<span
  title="Select serial port (Alt+P)">
  <div
    className="padded-row">
    <DropdownButton
      className="btn-primary btn-nordic"
      id="navbar-dropdown"
      onToggle={[Function]}
      open={true}
      title="Select serial port" />
    <div
      className="indicator off" />
  </div>
</span>
`;

exports[`AdapterSelector should render empty adapter list, not expanded 1`] = `
<span
  title="Select serial port (Alt+P)">
  <div
    className="padded-row">
    <DropdownButton
      className="btn-primary btn-nordic"
      id="navbar-dropdown"
      onToggle={[Function]}
      open={false}
      title="Select serial port" />
    <div
      className="indicator off" />
  </div>
</span>
`;

exports[`AdapterSelector should render empty list while loading 1`] = `
<span
  title="Select serial port (Alt+P)">
  <div
    className="padded-row">
    <DropdownButton
      className="btn-primary btn-nordic"
      id="navbar-dropdown"
      onToggle={[Function]}
      open={true}
      title="Select serial port" />
    <div
      className="indicator off" />
  </div>
</span>
`;

exports[`AdapterSelector should render two adapters 1`] = `
<span
  title="Select serial port (Alt+P)">
  <div
    className="padded-row">
    <DropdownButton
      className="btn-primary btn-nordic"
      id="navbar-dropdown"
      onToggle={[Function]}
      open={true}
      title="Select serial port">
      <MenuItem
        className="btn-primary"
        eventKey="/dev/tty0"
        onSelect={[Function]}>
        <div
          className="serialPort">
          /dev/tty0
        </div>
        <div
          className="serialSerialnumber">
          123456
        </div>
      </MenuItem>
      <MenuItem
        className="btn-primary"
        eventKey="/dev/tty1"
        onSelect={[Function]}>
        <div
          className="serialPort">
          /dev/tty1
        </div>
        <div
          className="serialSerialnumber">
          456789
        </div>
      </MenuItem>
    </DropdownButton>
    <div
      className="indicator off" />
  </div>
</span>
`;

exports[`AdapterSelector should render two adapters, one selected 1`] = `
<span
  title="Select serial port (Alt+P)">
  <div
    className="padded-row">
    <DropdownButton
      className="btn-primary btn-nordic"
      id="navbar-dropdown"
      onToggle={[Function]}
      open={true}
      title="/dev/tty1">
      <MenuItem
        className="btn-primary"
        eventKey="/dev/tty0"
        onSelect={[Function]}>
        <div
          className="serialPort">
          /dev/tty0
        </div>
        <div
          className="serialSerialnumber">
          123456
        </div>
      </MenuItem>
      <MenuItem
        className="btn-primary"
        eventKey="/dev/tty1"
        onSelect={[Function]}>
        <div
          className="serialPort">
          /dev/tty1
        </div>
        <div
          className="serialSerialnumber">
          456789
        </div>
      </MenuItem>
      <MenuItem
        className="btn-primary"
        eventKey="Close adapter"
        onSelect={[Function]}>
        <div
          className="serialPort">
          Close adapter
        </div>
        <div
          className="serialSerialnumber">
          
        </div>
      </MenuItem>
    </DropdownButton>
    <div
      className="indicator off" />
  </div>
</span>
`;
